### PR TITLE
Optimize GatherGrad for AMD GPU

### PR DESCRIPTION
**Pull request by weixingzhang**
Optimize GatherGrad for AMD GPU.

### CODE REVIEW

File: orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu [hunk 1]
```diff
@@ -70,6 +70,55 @@ __global__ void _GatherGradImpl(
   }
 }
 
+template <typename T, typename Tin>
+__global__ void _GatherGradImpl_new(
```
Comment (suffiank) [inline orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:74]: Maybe a more specific name to distinguish the special case?

File: orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu [hunk 2]
```diff
@@ -112,18 +161,33 @@ void GatherGradImpl(
...
       grad_data,
       output_data,
       num_indices,
       num_inputs,
-      param_itrs,
-      stride);
+      //param_itrs,
+      stride); 
+  } else {
+    dim3 block(GPU_WARP_SIZE, 4);
```
Comment (suffiank) [inline orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:179]: I wonder if the block and grid sizes should be adjusted for standard case too.

File: orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu [hunk 3]
```diff
@@ -161,24 +159,20 @@ void GatherGradImpl(
       original_indices.get(), original_indices_sorted.get(),
       num_indices));
 
+  dim3 block(GPU_WARP_SIZE, 4);
+  dim3 grid(CeilDiv(num_indices, 4), CeilDiv(stride, GridDim::maxElementsPerThread * GPU_WARP_SIZE));
```
Comment (suffiank) [inline orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:163]: Previously it was `2*ceil(stride/128)` and now it's `ceil(stride/256)`?

Comment (weixingzhang) [reply orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:163]: I changed the block size as well. Previously, it was block(GPU_WARP_SIZE/4, 4), which is a bit weird.

Comment (suffiank) [reply orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:163]: Oh, I see. I guess the perf is unaffected then?

Comment (weixingzhang) [reply orttraining/orttraining/training_ops/rocm/tensor/gather_grad_impl.cu:163]: yes, I measured and it is same.

**Review by suffiank — APPROVED** [review]

**Pull request merged**